src/main/resources/db/migration/R__functions.sql

```diff
@@ -0,0 +1,335 @@
+$body$
+BEGIN
+    -- NOTE This one is safe from race conditions only if
+    -- lock_versions is called before in the same transaction.
+    IF EXISTS(SELECT * FROM versions) THEN
+        -- create a new session-id+serial entry only in case there are
+        -- new entries in the log since the last session-id+serial.
+        -- (if there's anything new in the object_log)
+        WITH previous_version AS (
+            SELECT id, session_id, serial, last_log_entry_id
+            FROM versions v
+            ORDER BY id DESC
+            LIMIT 1
+        )
+        INSERT
+        INTO versions (session_id, serial, last_log_entry_id)
+        SELECT v.session_id, v.serial + 1, ol.id
+        FROM object_log ol,
+             previous_version v
+        WHERE ol.id > v.last_log_entry_id
+        ORDER BY ol.id DESC
+        LIMIT 1;
+    ELSEIF EXISTS(SELECT * FROM object_log) THEN
+        -- create a new version entry from object_log
+        -- (if there's anything new in the object_log)
+        INSERT INTO versions (session_id, serial, last_log_entry_id)
+        SELECT uuid_in(md5(random()::TEXT || clock_timestamp()::TEXT)::CSTRING), 1, id
+        FROM object_log
+        ORDER BY id DESC
+        LIMIT 1;
+    ELSE
+        -- this can happen on a completely empty database, not log, no versions
+        -- just initialise with something reasonable
+        INSERT INTO versions (session_id, serial, last_log_entry_id)
+        VALUES (uuid_in(md5(random()::TEXT || clock_timestamp()::TEXT)::CSTRING), 1, 0);
+
+    END IF;
+
+    RETURN QUERY
+        SELECT *
+        FROM versions
+        ORDER BY id DESC
+        LIMIT 1;
+END
+$body$
+    LANGUAGE plpgsql;
+
+
+
+-- Returns if there's any entry in the object_log table since
+-- the last frozen version.
+CREATE OR REPLACE FUNCTION changes_exist() RETURNS BOOLEAN AS
+$body$
+BEGIN
+    IF (EXISTS (SELECT * FROM versions)) THEN
+        RETURN EXISTS (
+            SELECT * FROM object_log
+            WHERE id > (
+                SELECT last_log_entry_id
+                FROM versions
+                ORDER BY id DESC
+                LIMIT 1)
+            );
+    ELSE
+        RETURN EXISTS (SELECT * FROM object_log);
+    END IF;
+END
+$body$
+    LANGUAGE plpgsql;
+
+
+-- Delete versions that
+-- a) have different session id than the latest version
+-- b) are not reasonable to keep, i.e. total size of deltas in the session
+-- becomes bigger than the snapshot size
+-- c) not the last version
+CREATE OR REPLACE FUNCTION delete_old_versions() RETURNS SETOF versions AS
+$body$
+BEGIN
+    --
+    RETURN QUERY
+        WITH latest_version AS (
+            SELECT *
+            FROM versions
+            ORDER BY id DESC
+            LIMIT 1
+        ),
+             deleted_versions AS (
+                 DELETE FROM versions
+                     WHERE id NOT IN (SELECT id FROM reasonable_deltas)
+                         AND id NOT IN (SELECT id FROM latest_version)
```

Comment:
Kinda suspect that not in latest version here already follows from the fact we only choose reasonable deltas, since reasonable deltas only take into account latest versions which has reasonable delta size sum.

Reply:
No, it breaks the tests if I remove the `AND id NOT IN (SELECT id FROM latest_version)` part.
Actually I added it there to fix the tests.